src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -310,6 +321,41 @@
     /** Actions to add to the {@link FlowStartNode}. */
     transient final List<Action> flowStartNodeActions = new ArrayList<Action>();
 
+    enum TimingKind {
+        /**
+         * Parsing Groovy sources; includes {@link #load}.
+         * @see CpsGroovyShell#parse(GroovyCodeSource)
+         */
+        parse,
+        /**
+         * Loading classes needed during {@link #parse}.
+         * @see ClassLoader#loadClass(String, boolean)
+         * @see ClassLoader#getResource
+         */
+        load,
+        /**
+         * Running inside {@link CpsVmExecutorService}, which includes many other things.
+         */
+        run,
+        /**
+         * Saving the program state.
+         * @see CpsThreadGroup#saveProgram(File)
+         */
+        save,
+        /**
+         * Loading or saving flow nodes.
+         * @see FlowNodeStorage
+         */
+        flow
+    }
+
+    /** accumulated time in ns of a given kind */
+    @GuardedBy("this")
+    @CheckForNull private Map</* for pretty XStream form */String, Long> timings;
+    /** last start time in ns of a given kind; unreliable to just subtract on start and add on end, since the VM might die in between */
+    @GuardedBy("this")
+    @CheckForNull private transient Map<TimingKind, Long> starts;
+
```

Comment:
🐜  but more of a trivial RFE: to make this useful, we should add one more map for call counts.  This lets us distinguish between a few very long calls vs. a lot of short ones. 

Without this we aren't likely to have the information needed to attack bottlenecks.  It will ALSO let us see if a method function simply has not been run enough times to JIT, which makes a huge difference.